web_widget_html_markdown/static/src/js/web_widget_html_markdown.js

```diff
@@ -0,0 +1,888 @@
+              self = this;
+              // This will prevent WYSIWYG from autodestructing if valid MD preservation
+              // tag is present
+              self.isSwitch = true;
+              var markdown_content = self.markdownEditor[0].value
+               // Take the markdown content wrap it and add it to html
+              var tag = self._generate_md_tag(markdown_content);
+
+              // Self.value =  self._getHtmlValue(self.value);
+              self.value =  self._getHtmlValue(markdown_content) + tag
+              // Remove editor and bars
+              self.markdownEditor.parent().remove();
+              self.$is_markdown = false;
+              self.markdownEditor = false
+              // Create new WYSIWYG in the process, WYSIWYG is creation is async
+              // BUG  At this point we must edit in HTML AGAIN TO NOT LOSE EDIT
+              return self._renderEdit()
+        },
+
+        sw_html_to_mk: function() {
+              self = this;
+              // markdown editor has problems with "falsy" values
+              // must be string because makeMarkdown function applies replace 
+              // function on src, without string conversion, causing error.
+              if (!self.value){self.value = "" }; 
+              var markdown_value = self._getMarkdownValue(self.value);
+              self._createMarkdownEditorInstance();
+              self.wysiwyg.destroy();
+              self.$is_markdown = true;
+              // Insert in markdown editor
+              self.markdownEditor.append(markdown_value)
+        },
+        start: function() {
+            this._super();
+            // This is custom how to add CSS class and styles
+            // Showdown does not support this out of the box.
+            // eslint-disable-next-line max-len
+            const classMap = {
+                table: "table table-striped",
+            };
+            var cmdItalic = null;
+
+            const clss_bindings = Object.keys(classMap).map(key => ({
+                type: "output",
+                regex: new RegExp(`<${key}(.*)>`, "g"),
+                replace: `<${key} class="${classMap[key]}" $1>`,
+            }));
+            this.$use_markdown = this.attrs.options.use_markdown || false;
+            this.$widget = this.attrs.widget;
+            this.$type = this.field.type;
+            // Inject MARKDOWN BUTTON
+            var md_s = $
+        (
+" <div id='radio_container'> <input id='markdown' type='radio' name='mk_html_switcher' value='markdown'>"+
+" <label id='markdown_label' for='markdown'>Markdown</label>  " +
+"<input type='radio' id='html' name='mk_html_switcher' value='html' checked >" +
+"  <label for='html'>HTML</label></div>")
+            var self = this;
+            this.$el.prepend(md_s);
+            //
+            this.$el.find('input').addClass('mk_html_switch');
+            this.$el.find('input').change([self], self._switch_markdown_html);
```

Comment:
I think it's best practice to make a separate sub-widget out of this and then instantiate it with:


```
this.sub_widget = new SubWidgetClass(this);
this.$el.prepend(this.sub_widget);
```

And then do the events in the Odoo `events:` way, let them bubble up to here via messages, and then catch them here.

Reply:
@thomaspaulb Question: Are we sure?  All other issues are done. I see the radiobutton+the editors as one widget.
The radiobutton+information_area as a separate widget would make sense if it was reusable elsewhere, this part of the widget is used once and makes sense only in conjunction with the rest. so I don't see why we should separate this.

After the cleanup, the widget is 200lines, i have refactored the views in XML, so if your concern is length of code it should not be a problem.  
I am all for modularity, but this sub-widget has not use anywhere else.